pkg/controller/oam/applicationconfiguration/applicationconfiguration.go

```diff
@@ -0,0 +1,315 @@
+			params:   ParameterResolveFn(resolve),
+			workload: ResourceRenderFn(renderWorkload),
+			trait:    ResourceRenderFn(renderTrait),
+		},
+		workloads: &workloads{
+			client: m.GetClient(),
+		},
+		gc:     GarbageCollectorFn(eligible),
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, ro := range o {
+		ro(r)
+	}
+
+	return r
+}
+
+// NOTE(negz): We don't validate anything against their definitions at the
+// controller level. We assume this will be done by validating admission
+// webhooks.
+
+// Reconcile an OAM ApplicationConfigurations by rendering and instantiating its
+// Components and Traits.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	ac := &v1alpha2.ApplicationConfiguration{}
+	if err := r.client.Get(ctx, req.NamespacedName, ac); err != nil {
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetAppConfig)
+	}
+
+	log = log.WithValues("uid", ac.GetUID(), "version", ac.GetResourceVersion())
+
+	workloads, err := r.components.Render(ctx, ac)
+	if err != nil {
+		log.Debug("Cannot render components", "error", err, "requeue-after", time.Now().Add(shortWait))
+		r.record.Event(ac, event.Warning(reasonCannotRenderComponents, err))
+		ac.Status.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errRenderComponents)))
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, ac), errUpdateAppConfigStatus)
+	}
+	log.Debug("Successfully rendered components", "workloads", len(workloads))
+	r.record.Event(ac, event.Normal(reasonRenderComponents, "Successfully rendered components", "workloads", strconv.Itoa(len(workloads))))
+
+	if err := r.workloads.Apply(ctx, workloads); err != nil {
+		log.Debug("Cannot apply components", "error", err, "requeue-after", time.Now().Add(shortWait))
+		r.record.Event(ac, event.Warning(reasonCannotApplyComponents, err))
+		ac.Status.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errApplyComponents)))
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, ac), errUpdateAppConfigStatus)
+	}
+	log.Debug("Successfully applied components", "workloads", len(workloads))
+	r.record.Event(ac, event.Normal(reasonApplyComponents, "Successfully applied components", "workloads", strconv.Itoa(len(workloads))))
+
+	// Kubernetes garbage collection will (by default) reap workloads and traits
+	// when the appconfig that controls them (in the controller reference sense)
+	// is deleted. Here we cover the case in which a component or one of its
+	// traits is removed from an extant appconfig.
+	for _, e := range r.gc.Eligible(ac.GetNamespace(), ac.Status.Workloads, workloads) {
+		// https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable
```

Comment:
Not sure why we need this.  AFAIK, the reference to e is only used once in an inline func call and not stored anymore.

Reply:
Yeah, this results in a false positive for https://github.com/kyoh86/scopelint, which we run as part of golangci-lint. In this case I chose to preventatively pin the variable rather than adding a `// nolint:scopelint` comment marker to disable the linter warning.